sklearn/utils/__init__.py

```diff
@@ -180,111 +181,53 @@ def axis0_safe_slice(X, mask, len_mask):
-
-    Returns
-    -------
-    subset
-        Subset of X on axis 0 or 1.
-
-    Notes
-    -----
-    CSR, CSC, and LIL sparse matrices are supported. COO sparse matrices are
-    not supported.
-    """
-    if axis == 0:
-        return _safe_indexing_row(X, indices)
-    elif axis == 1:
-        return _safe_indexing_column(X, indices)
-    else:
-        raise ValueError(
-            "'axis' should be either 0 (to index rows) or 1 (to index "
-            " column). Got {} instead.".format(axis)
-        )
-
-
-def _array_indexing(array, key, axis=0):
+def _array_indexing(array, key, axis):
     """Index an array consistently across NumPy version."""
-    if axis not in (0, 1):
-        raise ValueError(
-            "'axis' should be either 0 (to index rows) or 1 (to index "
-            " column). Got {} instead.".format(axis)
-        )
     if np_version < (1, 12) or issparse(array):
+        # FIXME: Remove the check for NumPy when using >= 1.12
         # check if we have an boolean array-likes to make the proper indexing
         key_array = np.asarray(key)
         if np.issubdtype(key_array.dtype, np.bool_):
             key = key_array
     return array[key] if axis == 0 else array[:, key]
 
 
-def _safe_indexing_row(X, indices):
-    """Return items or rows from X using indices.
-
-    Allows simple indexing of lists or arrays.
-
-    Parameters
-    ----------
-    X : array-like, sparse-matrix, list, pandas.DataFrame, pandas.Series
-        Data from which to sample rows or items.
-    indices : array-like of int
-        Indices according to which X will be subsampled.
-
-    Returns
-    -------
-    subset
-        Subset of X on first axis.
-
-    Notes
-    -----
-    CSR, CSC, and LIL sparse matrices are supported. COO sparse matrices are
-    not supported.
-    """
-    if hasattr(X, "iloc"):
-        # Work-around for indexing with read-only indices in pandas
-        indices = np.asarray(indices)
-        indices = indices if indices.flags.writeable else indices.copy()
-        # Pandas Dataframes and Series
-        try:
-            return X.iloc[indices]
-        except ValueError:
-            # Cython typed memoryviews internally used in pandas do not support
-            # readonly buffers.
-            warnings.warn("Copying input dataframe for slicing.",
-                          DataConversionWarning)
-            return X.copy().iloc[indices]
-    elif hasattr(X, "shape"):
-        if hasattr(X, 'take') and (hasattr(indices, 'dtype') and
-                                   indices.dtype.kind == 'i'):
-            # This is often substantially faster than X[indices]
-            return X.take(indices, axis=0)
-        else:
-            return _array_indexing(X, indices, axis=0)
+def _pandas_indexing(X, key, axis):
+    """Index a pandas dataframe or a series."""
+    # check whether we should index with loc or iloc
+    if _check_key_type(key, int):
+        by_name = False
```

Comment:
should we be raising a warning/error if `X.index.dtype.kind in 'iu'` (and similar on the other axis)? This won't handle the case where there are ints of object dtype in the index.